### PR TITLE
Add hicstuff full pipeline

### PR DESCRIPTION
**Pull request by loraine-gueguen**
The hicstuff `pipeline` subcommand, with a subset of parameters, has been implemented as a starting point.
Other hicstuff subcommands, missing parameters and output files can be implemented later.

FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

**Comment by loraine-gueguen** [comment]
local planemo test passing, but failing in CI, I don't know why

**Comment by loraine-gueguen** [comment]
Error: 
```
WARNING :: Using BSD sort instead of GNU sort, sorting on a single thread.
sort: unrecognized option '--temporary-directory=results/tmp'
```
`coreutils` package needed:
https://github.com/koszullab/hicstuff/blob/f0983469e56fa9c285faa1dbafb222035ba77027/hicstuff/io.py#L1210-L1226

File: tools/hicstuff/hicstuff_pipeline.xml [hunk 1]
```diff
@@ -0,0 +1,90 @@
+<tool id="hicstuff_pipeline" name="hicstuff full pipeline" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" python_template_version="3.5">
+    <description>generates a Hi-C contact matrix</description>
+    <macros>
+        <import>macros.xml</import>
+        <token name="@VERSION_SUFFIX@">0</token>
+    </macros>
+    <expand macro="requirements" />
+    <command detect_errors="exit_code"><![CDATA[
+hicstuff pipeline 
+    --genome $genome
```
Comment (bgruening) [inline tools/hicstuff/hicstuff_pipeline.xml:10]: single quotes missing

File: tools/hicstuff/macros.xml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+<macros>
+	<token name="@TOOL_VERSION@">3.1.5</token>
```
Comment (bgruening) [inline tools/hicstuff/macros.xml:2]: Indentation seems to be off here

File: tools/hicstuff/hicstuff_pipeline.xml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    --matfmt $matfmt
+    --quality-min $quality_min
+    --size $size
+    --threads \${GALAXY_SLOTS:-1}
+    $forward_reads
+    $reverse_reads
+    ]]></command>
+    <inputs>
+        <param type="data" name="genome" format="fasta" label="Genome fasta file"/>
+        <param type="data" name="forward_reads" format="fastqsanger" label="Forward reads"/>
```
Comment (bgruening) [inline tools/hicstuff/hicstuff_pipeline.xml:27]: This should be an advanced conditional where you can choose to put in a paired collection?

Comment (bernt-matthias) [reply tools/hicstuff/hicstuff_pipeline.xml:27]: Does the tool "understand" `fastqsanger.gz`?

File: tools/hicstuff/hicstuff_pipeline.xml (hunk 1 above)
Comment (loraine-gueguen) [reply tools/hicstuff/hicstuff_pipeline.xml]: done

File: tools/hicstuff/macros.xml (hunk 2 above)
Comment (loraine-gueguen) [reply tools/hicstuff/macros.xml]: done

File: tools/hicstuff/hicstuff_pipeline.xml (hunk 3 above)
Comment (loraine-gueguen) [reply tools/hicstuff/hicstuff_pipeline.xml]: > Does the tool "understand" `fastqsanger.gz`?

yes. Added

Comment (loraine-gueguen) [reply tools/hicstuff/hicstuff_pipeline.xml]: > This should be an advanced conditional where you can choose to put in a paired collection?

I can only find examples with paired reads and, for the full pipeline, starting with reads, 2 input files are required:
https://github.com/koszullab/hicstuff/blob/f0983469e56fa9c285faa1dbafb222035ba77027/hicstuff/pipeline.py#L592

Comment (bernt-matthias) [reply tools/hicstuff/hicstuff_pipeline.xml]: I guess @bgruening means the option to select fwd and reverse datasets separately (which you already implemented) and as a single paired dataset

This is usually done in a conditional. Like [here](https://github.com/galaxyproject/tools-iuc/blob/d9cb4478858f60aacb8577eb16e291961384eb1c/tools/dada2/macros.xml#L78) .. just remove the `"single"` option .. and you will need to adapt the command block (the paired dataset has attributes `.forward` and `.reverse` which refer to the actual datasets).

Comment (bernt-matthias) [reply tools/hicstuff/hicstuff_pipeline.xml]: More docs may be found [here](https://planemo.readthedocs.io/en/latest/writing_advanced.html#consuming-collections)

File: tools/hicstuff/hicstuff_pipeline.xml [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            </param>
+            <when value="paired">
+                <param name="reads" type="data_collection" collection_type="paired" format="fastqsanger,fastqsanger.gz" label="Paired reads"/>
+            </when>
+            <when value="separate">
+                <param name="forward_reads" type="data" format="fastqsanger,fastqsanger.gz" label="Forward reads"/>
+                <param name="reverse_reads" type="data" format="fastqsanger,fastqsanger.gz" label="Reverse reads"/>
+            </when>
+        </conditional>
+	    <param argument="--aligner" type="select" label="Alignment software to use" help="Minimap2 should only be used for reads > 100 bp">
```
Comment (bernt-matthias) [inline tools/hicstuff/hicstuff_pipeline.xml:45]: Excellent. The `conditional` looks good. 

Indentation is off here.

File: tools/hicstuff/hicstuff_pipeline.xml [hunk 5]
```diff
@@ -0,0 +1,106 @@
+<tool id="hicstuff_pipeline" name="hicstuff full pipeline" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" python_template_version="3.5">
```
Comment (bernt-matthias) [inline tools/hicstuff/hicstuff_pipeline.xml:1]:
```suggestion
<tool id="hicstuff_pipeline" name="hicstuff full pipeline" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@">
```

python_template_version is not needed anymore. profile is good to have.

File: tools/hicstuff/macros.xml [hunk 6]
```diff
@@ -0,0 +1,24 @@
+<macros>
+    <token name="@TOOL_VERSION@">3.1.5</token>
```
Comment (bernt-matthias) [inline tools/hicstuff/macros.xml:2]:
```suggestion
    <token name="@TOOL_VERSION@">3.1.5</token>
    <token name="@PROFILE@">21.09</token>
```

File: tools/hicstuff/macros.xml [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    <xml name="requirements">
+        <requirements>
+        <requirement type="package" version="3.1.5">hicstuff</requirement>
+            <yield/>
+        </requirements>
+    </xml>
+    <xml name="citations">
+        <citations>
+            <citation type="doi">10.5281/zenodo.4066363</citation>
+            <citation type="bibtex">
```
Comment (bernt-matthias) [inline tools/hicstuff/macros.xml:12]: Use the citation(s) given on the projects homepage: https://github.com/koszullab/hicstuff#citation

File: tools/hicstuff/.shed.yml [hunk 8]
```diff
@@ -0,0 +1,10 @@
+categories:
+- Sequence Analysis
+description: A toolkit to generate and manipulate Hi-C matrices
+homepage_url: https://github.com/koszullab/hicstuff
```
Comment (bernt-matthias) [inline tools/hicstuff/.shed.yml:4]: Its a bit odd that this is not the main fork, isn't it?

File: tools/hicstuff/hicstuff_pipeline.xml [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            </when>
+        </conditional>
+	    <param argument="--aligner" type="select" label="Alignment software to use" help="Minimap2 should only be used for reads > 100 bp">
+            <option value="bowtie2" selected="true">bowtie2</option>
+            <option value="minimap2">minimap2</option>
+            <option value="bwa">bwa</option>
+        </param>
+        <param argument="--circular" type="boolean" truevalue="--circular" falsevalue="" label="Circular genome"/>
+        <param argument="--duplicates" type="boolean" truevalue="--duplicates" falsevalue="" label="Removes PCR duplicates" help="PCR duplicates are defined as sets of pairs having identical mapping positions for both reads."/>
+        <param argument="--enzyme" type="text" value="5000" label="Bin size or enzyme" help="Restriction enzyme or 'mnase' if a string, or chunk size (i.e. resolution) if a number. Can also be multiple comma-separated enzymes."/>
```
Comment (bernt-matthias) [inline tools/hicstuff/hicstuff_pipeline.xml:52]: For user friendly interface you may think of a conditional here allowing to enter either an integer or select enzymes (I guess the list of enzymes is known). 

Up to you.

File: tools/hicstuff/.shed.yml [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+long_description: |
+  A toolkit to generate and manipulate Hi-C matrices
+name: hicstuff
```
Comment (bernt-matthias) [inline tools/hicstuff/.shed.yml:7]: I would suggest to create this as a suite. Its likely that more tools of the library will be added, or?

File: tools/hicstuff/.shed.yml (hunk 8 above)
Comment (loraine-gueguen) [reply tools/hicstuff/.shed.yml:4]: I agree, but this is the one used to build the python package on https://pypi.org/, used by the bioconda package

File: tools/hicstuff/macros.xml (hunk 6 above)
Comment (loraine-gueguen) [reply tools/hicstuff/macros.xml:2]: done

File: tools/hicstuff/macros.xml (hunk 7 above)
Comment (loraine-gueguen) [reply tools/hicstuff/macros.xml:13]: Done. I didn't notice that the fields were not filled...

File: tools/hicstuff/.shed.yml (hunk 10 above)
Comment (loraine-gueguen) [reply tools/hicstuff/.shed.yml]: Done (hopefully in the right way)

File: tools/hicstuff/hicstuff_pipeline.xml (hunk 9 above)
Comment (loraine-gueguen) [reply tools/hicstuff/hicstuff_pipeline.xml:52]: I would prefer to keep it that way for now, if you don't mind. I know it's not perfect and complete (other settings and subcommands are missing). The biohackathon coding time is over and I don't have much time to spend on it anymore...

But this a good starting point anyway!

File: tools/hicstuff/hicstuff_pipeline.xml (hunk 4 above)
Comment (loraine-gueguen) [reply tools/hicstuff/hicstuff_pipeline.xml]: Should be fixed

File: tools/hicstuff/hicstuff_pipeline.xml (hunk 9 above)
Comment (bernt-matthias) [reply tools/hicstuff/hicstuff_pipeline.xml:52]: Perfectly fine with me.

File: tools/hicstuff/.shed.yml [hunk 11]
```diff
@@ -0,0 +1,14 @@
+categories:
+- Sequence Analysis
+homepage_url: https://github.com/koszullab/hicstuff
+auto_tool_repositories:
+  name_template: "{{ tool_id }}"
+  description_template: "{{ tool_name }} (from the hicstuff tool suite)"
+suite:
+    name: hicstuff
```
Comment (bernt-matthias) [inline tools/hicstuff/.shed.yml:8]:
```suggestion
    name: suite_hicstuff
```

**Review by bernt-matthias — APPROVED** [review]
One last thing from my side. 

Thanks for the contribution.

**Comment by loraine-gueguen** [comment]
Thanks for your help @bernt-matthias !

**Comment by loraine-gueguen** [comment]
Hi @bernt-matthias @bgruening ! Could one of you please merge this PR ?

**Comment by bgruening** [comment]
Very cool, thanks!

**Pull request merged**